### PR TITLE
Fix bug #3150, regression introduced by a218453.

### PR DESCRIPTION
**Pull request by ezyang**
Added tests for this case, as well as some more.  This
probably breaks the Windows build.

Signed-off-by: Edward Z. Yang <ezyang@cs.stanford.edu>

### CODE REVIEW

**Comment by 23Skidoo** [comment]
> This probably breaks the Windows build.

You mean the whole build or just the test suite?

**Comment by ezyang** [comment]
Just the test suite :o)

**Comment by ezyang** [comment]
Speaking of which, can we get GitHub PRs to light up Appveyor builds?

**Comment by 23Skidoo** [comment]
Not sure whether it's supported.

**Comment by ezyang** [comment]
It should be possible; GitHub has an API for it: https://developer.github.com/v3/repos/statuses/ I guess we need to give Appveyor permissions to update the PRs?

**Comment by 23Skidoo** [comment]
I updated the webhook settings to deliver pull request notifications to AppVeyor.

**Comment by ezyang** [comment]
AppVeyor is running! Very cool.

**Comment by 23Skidoo** [comment]
Great!

**Comment by 23Skidoo** [comment]
LGTM, thanks for writing tests.

**Comment by ezyang** [comment]
Well! I wasn't expecting the SameDirectory tests to fail too. I guess 'exec' is no good on Windows. Wonder what a good substitute is.

File: Cabal/tests/PackageTests/GhcPkgGuess/SameDirectoryGhcVersion/ghc-7.10 [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
```
Comment (23Skidoo) [inline Cabal/tests/PackageTests/GhcPkgGuess/SameDirectoryGhcVersion/ghc-7.10]: BTW, looks like you forgot to include `ghc-*` scripts in `extra-source-files`.

**Comment by 23Skidoo** [comment]
Maybe we can work around it by adding `ghc-pkg-*.cmd` BAT scripts in addition to `ghc-pkg-*` shell scripts. Though our code right now doesn't recognise `cmd` as an exe extension.

**Comment by 23Skidoo** [comment]
And I guess the problem is also that the code looking for `ghc-pkg.exe`, not `ghc-pkg` on Windows.

**Comment by 23Skidoo** [comment]
OK, I think I'll just merge this now, and we'll fix the Windows error later.

**Comment by 23Skidoo** [comment]
Merged, thanks!

**Comment by grayjay** [comment]
We should probably filter the tests with something like `--pattern='!GhcPkgGuess'` until they can run on Windows.

**Comment by ezyang** [comment]
I was also willing to test a conditional before running them in the test suite itself, but I didn't see any prior way of saying "don't run this on Windows" in the test suite. Do we have a recommended way of doing this?

**Comment by grayjay** [comment]
I don't know.  I didn't consider that we would need to skip tests for other reasons when I made #3146, so it's not very flexible.

**Comment by ezyang** [comment]
It seems to me that it would be more scalable if we could detect what features are supported, and just adjust the test suite accordingly. (So if you're running Windows dev, you don't have to pass an extra flag to stop teh test suite from bleating.)

**Comment by 23Skidoo** [comment]
Disabled `GhcPkgGuess` tests on Windows.

**Comment by grayjay** [comment]
Thanks.

> It seems to me that it would be more scalable if we could detect what features are supported, and just adjust the test suite accordingly.

If we do this, I think it would be useful to have an option to enable all tests.  Then we could use it in combination with `--list-tests` to see what tests would be skipped in the current environment.  If tests check the environment, it would probably be easier to just revert #3146 and handle shared libraries similarly.

**Comment by dcoutts** [comment]
@ezyang thanks muchly! This had been annoying me for some time but not quite enough to fix it myself :-)

**Pull request merged**